File: backend/app/core/rbac/permissions.py
from enum import Enum
from typing import Dict, List, Set


class Permission(str, Enum):
    INCIDENTS_READ = "incidents:read"
    INCIDENTS_WRITE = "incidents:write"
    INCIDENTS_DELETE = "incidents:delete"

    REPORTS_READ = "reports:read"
    REPORTS_GENERATE = "reports:generate"

    ASSETS_READ = "assets:read"
    ASSETS_MANAGE = "assets:manage"

    USERS_READ = "users:read"
    USERS_MANAGE = "users:manage"

    PLAYBOOKS_READ = "playbooks:read"
    PLAYBOOKS_EXECUTE = "playbooks:execute"

    SETTINGS_MANAGE = "settings:manage"

    BILLING_READ = "billing:read"
    BILLING_MANAGE = "billing:manage"

    THREAT_INTEL_READ = "threat_intel:read"

    DASHBOARD_READ = "dashboard:read"
    AUDIT_READ = "audit:read"

    PLATFORM_ORGS_MANAGE = "platform:orgs:manage"
    PLATFORM_USERS_MANAGE = "platform:users:manage"
    PLATFORM_BILLING_MANAGE = "platform:billing:manage"
    PLATFORM_AUDIT_READ = "platform:audit:read"
    PLATFORM_SETTINGS_MANAGE = "platform:settings:manage"
    PLATFORM_HEALTH_VIEW = "platform:health:view"


PERMISSION_GROUPS: Dict[str, Set[str]] = {
    "incidents": {
        Permission.INCIDENTS_READ,
        Permission.INCIDENTS_WRITE,
        Permission.INCIDENTS_DELETE,
    },
    "reports": {
        Permission.REPORTS_READ,
        Permission.REPORTS_GENERATE,
    },
    "assets": {
        Permission.ASSETS_READ,
        Permission.ASSETS_MANAGE,
    },
    "users": {
        Permission.USERS_READ,
        Permission.USERS_MANAGE,
    },
    "playbooks": {
        Permission.PLAYBOOKS_READ,
        Permission.PLAYBOOKS_EXECUTE,
    },
    "settings": {
        Permission.SETTINGS_MANAGE,
    },
    "billing": {
        Permission.BILLING_READ,
        Permission.BILLING_MANAGE,
    },
    "threat_intel": {
        Permission.THREAT_INTEL_READ,
    },
    "dashboard": {
        Permission.DASHBOARD_READ,
    },
    "audit": {
        Permission.AUDIT_READ,
    },
    "platform": {
        Permission.PLATFORM_ORGS_MANAGE,
        Permission.PLATFORM_USERS_MANAGE,
        Permission.PLATFORM_BILLING_MANAGE,
        Permission.PLATFORM_AUDIT_READ,
        Permission.PLATFORM_SETTINGS_MANAGE,
        Permission.PLATFORM_HEALTH_VIEW,
    },
}


ROLE_PERMISSIONS: Dict[str, List[str]] = {
    "SUPER_ADMIN": [p.value for p in Permission],
    "ORG_ADMIN": [
        Permission.INCIDENTS_READ,
        Permission.INCIDENTS_WRITE,
        Permission.INCIDENTS_DELETE,
        Permission.REPORTS_READ,
        Permission.REPORTS_GENERATE,
        Permission.ASSETS_READ,
        Permission.ASSETS_MANAGE,
        Permission.USERS_READ,
        Permission.USERS_MANAGE,
        Permission.PLAYBOOKS_READ,
        Permission.PLAYBOOKS_EXECUTE,
        Permission.SETTINGS_MANAGE,
        Permission.BILLING_READ,
        Permission.BILLING_MANAGE,
        Permission.THREAT_INTEL_READ,
        Permission.DASHBOARD_READ,
        Permission.AUDIT_READ,
    ],
    "ANALYST": [
        Permission.INCIDENTS_READ,
        Permission.INCIDENTS_WRITE,
        Permission.REPORTS_READ,
        Permission.REPORTS_GENERATE,
        Permission.ASSETS_READ,
        Permission.PLAYBOOKS_READ,
        Permission.PLAYBOOKS_EXECUTE,
        Permission.THREAT_INTEL_READ,
    ],
}


def get_permissions_for_role(role: str) -> List[str]:
    return ROLE_PERMISSIONS.get(role, [])


def role_has_permission(role: str, permission: str) -> bool:
    if role == "SUPER_ADMIN":
        return True
    return permission in ROLE_PERMISSIONS.get(role, [])


def get_all_permissions() -> List[dict]:
    result = []
    for perm in Permission:
        group = next((g for g, perms in PERMISSION_GROUPS.items() if perm.value in perms), "other")
        result.append({
            "permission": perm.value,
            "group": group,
            "description": perm.value.replace(":", " ").replace("_", " ").title(),
        })
    return result

File: frontend/src/lib/rbac/permissions.ts
export type Role = "SUPER_ADMIN" | "ORG_ADMIN" | "ANALYST";

export type Permission =
  | "incidents:read"
  | "incidents:write"
  | "incidents:delete"
  | "reports:read"
  | "reports:generate"
  | "assets:read"
  | "assets:manage"
  | "users:read"
  | "users:manage"
  | "playbooks:read"
  | "playbooks:execute"
  | "settings:manage"
  | "billing:read"
  | "billing:manage"
  | "threat_intel:read"
  | "platform:orgs:manage"
  | "platform:users:manage"
  | "platform:billing:manage"
  | "platform:audit:read"
  | "platform:settings:manage";

export const ROLE_PERMISSIONS: Record<Role, Permission[]> = {
  SUPER_ADMIN: [
    "incidents:read", "incidents:write", "incidents:delete",
    "reports:read", "reports:generate",
    "assets:read", "assets:manage",
    "users:read", "users:manage",
    "playbooks:read", "playbooks:execute",
    "settings:manage",
    "billing:read", "billing:manage",
    "threat_intel:read",
    "platform:orgs:manage", "platform:users:manage",
    "platform:billing:manage", "platform:audit:read", "platform:settings:manage",
  ],
  ORG_ADMIN: [
    "incidents:read", "incidents:write", "incidents:delete",
    "reports:read", "reports:generate",
    "assets:read", "assets:manage",
    "users:read", "users:manage",
    "playbooks:read", "playbooks:execute",
    "settings:manage",
    "billing:read", "billing:manage",
    "threat_intel:read",
  ],
  ANALYST: [
    "incidents:read", "incidents:write",
    "reports:read", "reports:generate",
    "assets:read",
    "playbooks:read", "playbooks:execute",
    "threat_intel:read",
  ],
};

export const PERMISSION_GROUPS: Record<string, Permission[]> = {
  incidents: ["incidents:read", "incidents:write", "incidents:delete"],
  reports: ["reports:read", "reports:generate"],
  assets: ["assets:read", "assets:manage"],
  users: ["users:read", "users:manage"],
  playbooks: ["playbooks:read", "playbooks:execute"],
  settings: ["settings:manage"],
  billing: ["billing:read", "billing:manage"],
  threat_intel: ["threat_intel:read"],
  platform: [
    "platform:orgs:manage",
    "platform:users:manage",
    "platform:billing:manage",
    "platform:audit:read",
    "platform:settings:manage",
  ],
};

export function roleHasPermission(role: Role, permission: Permission): boolean {
  if (role === "SUPER_ADMIN") return true;
  return ROLE_PERMISSIONS[role]?.includes(permission) ?? false;
}

export function getPermissionsForRole(role: Role): Permission[] {
  return ROLE_PERMISSIONS[role] ?? [];
}

export function getAllPermissions(): { permission: Permission; group: string; description: string }[] {
  const result: { permission: Permission; group: string; description: string }[] = [];
  for (const [group, permissions] of Object.entries(PERMISSION_GROUPS)) {
    for (const permission of permissions) {
      result.push({
        permission,
        group,
        description: permission.replace(":", " ").replace("_", " ").replace(/\b\w/g, l => l.toUpperCase()),
      });
    }
  }
  return result;
}